src/main/java/engineering/swat/watch/impl/jdk/JDKFileTreeWatch.java

```diff
@@ -0,0 +1,221 @@
+    private final Path rootPath;
+    private final Path relativePathParent;
+    private final Map<Path, JDKFileTreeWatch> childWatches = new ConcurrentHashMap<>();
+    private final JDKBaseWatch internal;
+
+    public JDKFileTreeWatch(Path fullPath, Executor exec,
+            BiConsumer<EventHandlingWatch, WatchEvent> eventHandler) {
+        this(fullPath, Path.of(""), exec, eventHandler);
+    }
+
+    public JDKFileTreeWatch(Path rootPath, Path relativePathParent, Executor exec,
+            BiConsumer<EventHandlingWatch, WatchEvent> eventHandler) {
+
+        super(rootPath.resolve(relativePathParent), exec, eventHandler);
+        this.rootPath = rootPath;
+        this.relativePathParent = relativePathParent;
+
+        var internalEventHandler = eventHandler.andThen(new ChildWatchesUpdater());
+        this.internal = new JDKDirectoryWatch(path, exec, internalEventHandler) {
+
+            // Override to ensure that this watch relativizes events wrt
+            // `rootPath` (instead of `path`, as is the default behavior)
+            @Override
+            public WatchEvent relativize(WatchEvent event) {
+                return new WatchEvent(event.getKind(), rootPath,
+                    rootPath.relativize(event.calculateFullPath()));
+            }
+
+            // Override to ensure that this watch translates JDK events using
+            // `rootPath` (instead of `path`, as is the default behavior).
+            // Events returned by this method do not need to be relativized.
+            @Override
+            protected WatchEvent translate(java.nio.file.WatchEvent<?> jdkEvent) {
+                var kind = translate(jdkEvent.kind());
+
+                Path relativePath = null;
+                if (kind != WatchEvent.Kind.OVERFLOW) {
+                    var child = (Path) jdkEvent.context();
+                    if (child != null) {
+                        relativePath = relativePathParent.resolve(child);
+                    }
+                }
+
+                var event = new WatchEvent(kind, rootPath, relativePath);
+                logger.trace("Translated: {} to {}", jdkEvent, event);
+                return event;
+            }
+        };
+    }
+
+    /**
+     * Event handler that updates the child watches according to the following
+     * rules: (a) when an overflow happens, it's propagated to each existing
+     * child watch; (b) when a subdirectory creation happens, a new child watch
+     * is opened for that subdirectory; (c) when a subdirectory deletion
+     * happens, an existing child watch is closed for that subdirectory.
+     */
+    private class ChildWatchesUpdater implements BiConsumer<EventHandlingWatch, WatchEvent> {
+        @Override
+        public void accept(EventHandlingWatch watch, WatchEvent event) {
+            switch (event.getKind()) {
+                case OVERFLOW: acceptOverflow(); break;
+                case CREATED: acceptCreated(event.calculateFullPath()); break;
```

Comment:
For most places I think we should avoid calling `calculateFullPath` it should be more of a down-stream user function to call. It might be needed here, but since we have a current root path, we could just resolve it ourselves? which also avoids future relativize calls?